### PR TITLE
Add S3Cache

### PR DESCRIPTION
**Pull request by alexjurkiewicz**
Is there interest in merging a cache backed with S3? If so, I will complete work on this PR.

At the moment it's incomplete:

- [x] Custom bucket
- [x] Key prefix
- [x] `has`, `get`, `set`
- [x] `add`
- [x] `delete`
- [x] `clear`
- [x] timeout
- [x] `delete_many`
- [ ] `inc`, `dec`
- [x] `get_dict`, `get_many`
- [x] `set_many`
- [x] Test

### CODE REVIEW

**Comment by ThiefMaster** [comment]
Not sure if that makes much sense - a cache is usually meant to be something very fast, and I don't think S3 is a great choice for that. AFAIK it's more something people use to put bigger files etc, but not the kind of data snippets one usually puts in a cache..

**Comment by alexjurkiewicz** [comment]
The performance of S3 is probably better than you expect. Keys can handle
thousands of reads and writes per second. Time to first byte is 100-200ms.
Both these data points are from here:
https://docs.aws.amazon.com/AmazonS3/latest/dev/optimizing-performance.html

S3 is in many ways similar to the filesystem cache. It would be useful in
situations where objects are relatively large and the cache miss time is
relatively high. For example, systems which deal with converting media
files, such as on-demand image resizing or audio re-encoding from master
copies.

S3 works in places where the filesystem cache does not, for example in AWS
compute resources. Most particularly Lambda, but also EC2/ECS/EKS, where
the filesystem cache could only work via EFS.

**Comment by ThiefMaster** [comment]
ok fair enough, for large cached objects it makes more sense

**Comment by alexjurkiewicz** [comment]
This is ready for review. Caveats / differences from base cache implementation:

1. I didn't implement `inc`/`dec` because it seemed pointless
2. TTL is implemented by setting the Expires header. I added a parameter to automatically clean up stale data, but real production use should probably rely on a S3 bucket lifecycle rule to clean up old objects.

I've added pytest for S3Cache but it's not hooked up to tox or Travis or what have you. I don't think that's in scope for this PR. You can run the tests with `python -m pytest`.

**Comment by alexjurkiewicz** [comment]
Also, this PR only works on Python 3. I assume that's fine? Also, it would be nice to bump the version of this package to 1.0.0 before merging this PR, and merge it as 1.1.0 (so the version numbers are useful).

**Comment by alexjurkiewicz** [comment]
Finally, testing cache expiry could be sped up with https://github.com/spulec/freezegun (I won't do it as part of this PR, this is just a note for the future).

**Comment by bfaludi** [comment]
@ThiefMaster @alexjurkiewicz Wondering would it be possible to merge this? :) 

Superset's documentation is mentioning an `s3cache` library but that package is not compatible with the werkzeug's abstract class anymore. Since most of the caching methods were moved here, It would be awesome to have S3 cache option back without hacking around this functionality. :)

https://superset.incubator.apache.org/installation.html?highlight=cache

File: cachelib/s3.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
+# -*- coding: utf-8 -*-
+import datetime
+
+try:
+    import cPickle as pickle
+except ImportError:  # pragma: no cover
+    import pickle
+
+try:
```
Comment (RonnyPfannschmidt) [inline cachelib/s3.py]: this bit has to go in the `__init__`, else the complete package fails without boto3 installed

**Comment by alexjurkiewicz** [comment]
Sorry, someone else will have to pick up the torch here, I'm not interested
in working further on it.

**Comment by lepture** [comment]
I don't think S3 is a good choice for cache.

**Pull request closed without merging**